### ISSUE TITLE
Does std::function immediately end lifetime in-place stored callable?

### ISSUE THREAD

**Issue by tomasz-kaminski-sonarsource** [microsoft/STL#3156]
With @balazs-benics-sonarsource, we  were recently looking at the report for the static analyzer, where the value of the `file` (`FILE*`) is considered to be leaked immediately after the following constructor on MSVC implementation:
```
std::function<void(f)> f([file] { close(file) })
```

After investigating the issue further and checking the corresponding wording, I think that in this particular situation, the `function` implementation may (per abstract machine rules) end the lifetime of the callable stored in place. Note that this is mostly a theoretical question and may not impact the compiled code.

We are particularly interested in:
[functional#L985-L999](https://github.com/microsoft/STL/blob/main/stl/inc/functional#L985-L999)
```
    union _Storage { // storage for small objects (basic_string is small)
        max_align_t _Dummy1; // for maximum alignment
        char _Dummy2[_Space_size]; // to permit aliasing _Space_size = (_Small_object_num_ptrs - 1) * sizeof(void*);
        _Ptrt* _Ptrs[_Small_object_num_ptrs]; // _Ptrs[_Small_object_num_ptrs - 1] is reserved
    };

    _Storage _Mystorage;
    enum { _EEN_IMPL = _Small_object_num_ptrs - 1 }; // helper for expression evaluator
    _Ptrt* _Getimpl() const noexcept { // get pointer to object
        return _Mystorage._Ptrs[_Small_object_num_ptrs - 1];
    }

    void _Set(_Ptrt* _Ptr) noexcept { // store pointer to object
        _Mystorage._Ptrs[_Small_object_num_ptrs - 1] = _Ptr;
    }
```
And `_Reset` that is called by constructor:
[functional#L909-L923](https://github.com/microsoft/STL/blob/main/stl/inc/functional#L909-L923)
````
    template <class _Fx>
    void _Reset(_Fx&& _Val) { // store copy of _Val
        if (!_Test_callable(_Val)) { // null member pointer/function pointer/std::function
            return; // already empty
        }

        using _Impl = _Func_impl_no_alloc<decay_t<_Fx>, _Ret, _Types...>;
        if constexpr (_Is_large<_Impl>) {
            // dynamically allocate _Val
            _Set(_Global_new<_Impl>(_STD forward<_Fx>(_Val)));
        } else {
            // store _Val in-situ
            _Set(::new (static_cast<void*>(&_Mystorage)) _Impl(_STD forward<_Fx>(_Val)));
        }
    }
````

In case when the in-situ path we create an `_Impl` object in the bytes on the `_Mystorage` (fully correct) and then invoke the `_Set` method that, in turn, stores the pointer in the object to `_Mystorage._Ptrs[_Small_object_num_ptrs - 1]`. The bytes of the pointer are guaranteed not to overlap with the bytes of the object.

However, per [[class.union] p6](https://eel.is/c++draft/class.union#general-6), the `_Mystorage._Ptrs[_Small_object_num_ptrs - 1]` creates whole `_Mystorage._Ptrs` reusing the storage  of `_Impl`. In consequence [[basic.life] p1.5](https://eel.is/c++draft/basic.life#1.5) the lifetime of the `_Impl` (and callable for which it provides storage) ends. Ending the lifetime is not necessarily leading to UB [[basic.life] p5](https://eel.is/c++draft/basic.life#5), but the object can no longer be accessed at this point without UB.

Do you agree with the above reasoning? If so, is there a particular reason to store `_Impl` pointer inside the union instead of keeping it a separate member that would not encounter the above issue?

**Comment by frederick-vs-ja** [comment]
I think your analysis makes sense, and there are several other places in the implementation of `std::function` raising UB. Probably `to permit aliasing` in the comment is wrong according to the standard, but it might be true for `/clr` or `/clr:pure`.

IMO we should just use an `unsigned char[_Space_size]` buffer that is `alignas(max_align_t)` to implement SFO and vptr, but it seems not suitable until vNext due to `/clr:pure`.

**Comment by AlexGuteniev** [comment]
We can fix `move_only_function` separately, though (as no ABI promises for it yet, and no /clr:pure need). Is it worth doing?

**Comment by StephanTLavavej** [comment]
We talked about this at the weekly maintainer meeting - our STL implementation is assuming (reasonable) behavior from our supported codegen compilers (MSVC and Clang). Although this is theoretically an issue according to the rules of the abstract machine, in practice there is no issue, and library implementations often need to make implementation assumptions beyond what the Standard provides. We recommend adjusting third-party static analysis tools to ignore such scenarios.

@frederick-vs-ja @AlexGuteniev We don't believe that any changes to `function` or `move_only_function` should be necessary here, unless and until they become `constexpr` (in which case we'd need to follow the abstract rules more strictly, as we saw with `string` and `vector`).
